Add FlowMenu and enhance console flow handling

abdelaziz-mahdy opened the pull request:
Introduce FlowMenu to the menu system for improved flow integration and update Console.js to manage 'readonly' flow mode in action availability and toggle functionality. Filter out help commands from flow memento before saving.

abdelaziz-mahdy committed:
Add FlowMenu and update Console flow handling

- Introduced FlowMenu to the menu system, allowing for flow integration.
- Updated Console.js to handle 'readonly' flow mode in action availability and toggle functionality.
- Adjusted menu registration to include FlowMenu in the main menu structure.

abdelaziz-mahdy committed:
Filter out help commands from flow memento before saving in Commands.js

abdelaziz-mahdy committed:
Remove filtering of help commands from flow memento before saving in Commands.js

kgrgreer commented on src/foam/core/menu/FlowMenu.js:
```diff
@@ -0,0 +1,34 @@
+      var view = {
+        class: 'foam.core.console.Console',
+        flowName: this.flow,
+        flowMode: 'readonly'
```
While you're working in this area, how about making this an Enum?

kgrgreer commented on src/foam/core/console/Console.js:
```diff
@@ -774,7 +779,11 @@ foam.CLASS({
-      code: function() { console.log('***', this.flowMode); this.flowMode = { edit: 'view', view: 'console', console: 'edit' }[this.flowMode]; },
+      code: function() { 
+        if (this.flowMode != 'readonly') {
+        console.log('***', this.flowMode); this.flowMode = { edit: 'view', view: 'console', console: 'edit' }[this.flowMode]; 
```
remove log()

kgrgreer commented on src/foam/core/console/Console.js:
```diff
@@ -774,7 +779,11 @@ foam.CLASS({
       // You can do this.showInput = true|false; from flow scripts
-      code: function() { console.log('***', this.flowMode); this.flowMode = { edit: 'view', view: 'console', console: 'edit' }[this.flowMode]; },
+      code: function() { 
+        if (this.flowMode != 'readonly') {
```
spacing

kgrgreer commented on src/foam/core/console/cmd/Commands.js:
```diff
@@ -533,6 +533,7 @@ foam.CLASS({
       // Don't save the 'save' command
       this.currentBlock.del();
 
+
```
Why the extra line?

abdelaziz-mahdy replied:
i was removing the helps on save, but i think we need to update it so readonly doesnt add any blocks. (maybe disable hyperline rendering on readonly?)

abdelaziz-mahdy replied:
if this is not important we can ignore it for now

kgrgreer commented on src/foam/core/menu/FlowMenu.js:
```diff
@@ -0,0 +1,34 @@
+  extends: 'foam.core.menu.AbstractMenu',
+
+  documentation: 'A menu item which contains a flow.',
+
```
remove extra line.

abdelaziz-mahdy committed:
Add FlowMode enum to Console and update flow handling

- Introduced FlowMode enum to define flow states: edit, view, readonly, and console.
- Updated Console.js to utilize FlowMode for flowMode property and related expressions.
- Adjusted FlowMenu to reference FlowMode for consistency in flow state management.

abdelaziz-mahdy committed:
Remove unnecessary blank line in Commands.js to improve code readability.

kgrgreer commented on src/foam/core/console/Console.js:
```diff
@@ -304,6 +304,11 @@ foam.CLASS({
+foam.ENUM({
+  package: 'foam.core.console',
+  name: 'FlowMode',
+  values: [ 'edit', 'view', 'readonly', 'console' ]
```
Enum values are CONSTANT by convention. ie. EDIT, VIEW, RO, CONSOLE

kgrgreer commented on src/foam/core/console/Console.js:
```diff
@@ -405,15 +411,17 @@ foam.CLASS({
       value: true,
       expression: function(flowMode) {
-        return flowMode == 'edit';
+        return flowMode == this.FlowMode.EDIT;
```
=== is better in this case

kgrgreer commented on src/foam/core/menu/FlowMenu.js:
```diff
@@ -0,0 +1,37 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
```
2025

kgrgreer commented on src/foam/core/menu/FlowMenu.js:
```diff
@@ -0,0 +1,37 @@
+
+      return view;
+     }
+  ],
```
no trailing ,

abdelaziz-mahdy committed:
Update FlowMode enum values to uppercase and improve strict equality checks in Console.js; update copyright year in FlowMenu.js

abdelaziz-mahdy committed:
Enhance flow validation in Commands.js

- Added checks to ensure a flow is available before execution.
- Implemented validation to require a name for the flow, providing user feedback if not set.

abdelaziz-mahdy committed:
Merge branch 'development' of https://github.com/kgrgreer/foam3 into reflow-menu

abdelaziz-mahdy committed:
Refactor flowMode checks in Console.js for consistency; update FlowMenu.js to use abbreviated flowMode value

kgrgreer commented on src/foam/core/console/cmd/Commands.js:
```diff
@@ -526,10 +526,22 @@ foam.CLASS({
   methods: [
     function execute(opt_flowName) {
+      // Check if flow is available here
+      if ( ! this.flow ) {
```
When does this happen?

abdelaziz-mahdy replied:
it doesnt happen, I forgot it while testing

kgrgreer commented on src/foam/core/console/cmd/Commands.js:
```diff
@@ -526,10 +526,22 @@ foam.CLASS({
         this.flow.name = opt_flowName;
       }
 
+      if ( ! this.flow.name || this.flow.name === 'Unnamed' ) {
```
Better if we call flow.save() since it already has this logic, then we don't have to duplicate it and possibly change it in two spots.
You can tell if it refuses to save() if it returns undefined.

abdelaziz-mahdy replied:
done

abdelaziz-mahdy committed:
Refactor flow validation in Commands.js to streamline checks for flow name and availability. Simplified logic by removing redundant checks and ensuring user feedback is provided when flow name is not set.
